File: gwas/wrangle_CB_gwas_cohort.R
# ## Phenotype data wrangling script
#
# Purpose: Wrangles data exported from Cohort Browser.
# Performs:
# - Assigns cases to participants with BREAST cancer type 
# - Assigns control to participants with no recorded cancer type 
# - Creates a matched case-control cohort (1000 cases and 1000 controls)
# - Derives age at recruitment

suppressMessages(library(tidyverse))

input_file = "../../mounted-data/female_cohort.csv"
female_cohort = readr::read_csv(input_file, na = c('', 'NA'))

# Remove rows without platekeys
t2 = female_cohort %>% dplyr::filter(!is.na(`Platekey-0.0`)) %>% unite( contains('Cancer_Disease_Type'), col = 'CANCERS', sep = ',')
t3 <- separate(data = t2, col = `Date_Of_Consent-0.0`, into = c("year_of_consent", "month_of_consent","day_of_consent"), sep = "-")
t3$year_of_consent <- as.numeric(t3$year_of_consent)
t3$`Year_Of_Birth-0.0` <- as.numeric(t3$`Year_Of_Birth-0.0`)
t4 = t3 %>% mutate( breast_cancer = ifelse(grepl("BREAST", CANCERS), 1, 0), age=as.numeric(year_of_consent)-as.numeric(`Year_Of_Birth-0.0`))
t5 = t4 %>% select(FID='Platekey-0.0', IID='Platekey-0.0', age, breast_cancer)

write_tsv(file="../../gwas_cohort_pheno_covariates.tsv", x=t5)

File: gwas/wrangle_CB_gwas_cohort.py
# ## Phenotype data wrangling script
#
# Purpose: Wrangles data exported from Cohort Browser.
# Performs:
# - Assigns cases to participants with BREAST cancer type 
# - Assigns control to participants with no recorded cancer type 
# - Creates a matched case-control cohort (1000 cases and 1000 controls)
# - Derives age at recruitment

import pandas as pd
pd.options.mode.chained_assignment = None

female_cohort = pd.read_csv('../../mounted-data/female_cohort.csv')
female_cohort.head()


def create_gwas_bc_cohort(phenotype_df,n_cases_controls_to_match=1000):
    phenotype_df = phenotype_df[~female_cohort['Platekey-0.0'].isna()]
    bc_cases = phenotype_df[(phenotype_df['Cancer_Disease_Type-0.0']=='BREAST') | (phenotype_df['Cancer_Disease_Type-0.1']=='BREAST') | (phenotype_df['Cancer_Disease_Type-0.2']=='BREAST')]
    bc_cases['breast_cancer'] = 1
    controls = phenotype_df[(phenotype_df['Cancer_Disease_Type-0.0'].isna()) & (phenotype_df['Cancer_Disease_Type-0.1'].isna()) & (phenotype_df['Cancer_Disease_Type-0.2'].isna())]
    controls['breast_cancer'] = 0
    cc_cohort = pd.concat([bc_cases.head(n_cases_controls_to_match), controls.head(n_cases_controls_to_match)])
    cc_cohort.columns = cc_cohort.columns.str.split('-').str[0]
    cc_cohort['year_of_consent'] = cc_cohort['Date_Of_Consent'].str.split('-').str[0]
    cc_cohort['year_of_consent'] = cc_cohort['year_of_consent'].astype(int)
    cc_cohort['age'] = cc_cohort['year_of_consent'] - cc_cohort['Year_Of_Birth']
    cc_cohort = cc_cohort.rename(columns={'Platekey':'IID'})
    cc_cohort['FID'] = cc_cohort['IID']
    cc_cohort = cc_cohort[['FID','IID','age','breast_cancer']]
    
    return cc_cohort


df = create_gwas_bc_cohort(female_cohort)

df.head()

df.to_csv('../../gwas_cohort_pheno_covariates.tsv',sep='\t',index=False)
